UTR Analysis Pipeline Test Dataset
==================================

## Overview

The purpose of this test dataset is to capture all possible scenarios for
which we might expect to find spliced leader (SL) sequences and Poly(A) tails
in RNA-Seq reads, and to differentiate those from similar, but non-real (false
positive) scenarios.

Using this test dataset, we can ensure that the pipeline is properly detecting
and characterizing all "true" SL/Poly(A) hits, and also ensure that it is
ignoring genomic sequences that _appear_ to be SL or Poly(A) derrived, but are
not.

Moreover, since different file formats and programming languages use different
coordinate systems (0- or 1-indexed), having a set of known sites with manually
determined coordinates helps to ensure that the code does not suffer from
off-by-one errors when reporting locations of SL and Poly(A) sites.

Each of the reads included comes from a real _T. cruzi_ Y strain RNA-Seq
sample. For speed and simplicity, we have restricted ourselves to the first
two chromosomes of *T. cruzi*.

The FASTQ file included here contains one read for each of the following
scenarios:

## Spliced leader

### Positive strand

- Portion of the SL at the start of read 
    1. HWI-1KL118:75:C267FACXX:4:1311:21375:97418 (R1, TcCLB.506717.130)
    2. HWI-1KL118:75:C267FACXX:4:2311:5511:72243 (R2, TcCLB.506717.219)
- Portion of the SL in the middle of the read, with artifactual (non-genomic)
   sequence upstream of it.
    3. HWI-1KL118:75:C267FACXX:4:2102:3429:91194 (R1, TcCLB.506533.106)
    4. HWI-1KL118:75:C267FACXX:4:2313:2265:3900 (R1, TcCLB.506533.106)
- Portion of the SL in a read, with a smaller genomic (non-real) hit in the
   middle of the read.

### Negative strand

- Portion of the SL at the start of read 
    - HWI-1KL118:75:C267FACXX:4:1315:20505:74568 (R1, TcCLB.507941.160)
    - HWI-1KL118:75:C267FACXX:4:1113:3259:52771 (R2, TcCLB.510351.50)
- Portion of the SL in the middle of the read, with artifactual (non-genomic)
   sequence upstream of it.
    - HWI-1KL118:75:C267FACXX:4:1113:11382:45693 (R1, TcCLB.507941.80)
    - HWI-1KL118:75:C267FACXX:4:1301:12715:37541 (R2, TcCLB.507941.40)
- Portion of the SL in a read, with a smaller genomic (non-real) hit in the
   middle of the read.

### Reverse complement

- Portion of the SL reverse complement at the start of a read
- Portion of the SL reverse complement in the middle of a read

## Poly(A) tail

1. HWI-1KL118:75:C267FACXX:4:1306:21290:65891 (R1)
2. HWI-1KL118:75:C267FACXX:4:2114:10637:92339 (R2)

### Reverse complement (Poly(T))

1. HWI-1KL118:75:C267FACXX:4:2306:2785:85962 (R1)
2. HWI-1KL118:75:C267FACXX:4:2102:4524:17908 (R2)

## Generation of test data

Test data was generated by extracting actual reads matching each of the
scenarios of interest from a collection of *T. cruzi* RNA-Seq samples.

In order to keep the test data as simple as possible, reads were limited
to a single sample (HPGL0254)

### x. Generating fastq files from list of read identifiers

First, appropriate test reads should be found for each of the above scenarios.
To do this, one could either grep through the original read fastq files to find
reads containing some large fraction of the SL sequence (and therefor likely to
actually include the spliced leader), although these need to be verified to
ensure that they are not arising from SL-like genomic sequences.

For example:

```
# 34-character suffix of T. cruzi SL sequence
zgrep -B1 -A2 ACGCTATTATTGATACAGTTTCTGTACTATATTG HPGL0254_genomic_reads_removed.1.fastq.gz

# 34-character suffix of T. cruzi SL sequence (reverse complement)
zgrep -B1 -A2 CAATATAGTACAGAAACTGTATCAATAATAGCGT HPGL0254_genomic_reads_removed.1.fastq.gz
```

If the test reads are described in this README, then the following grep
commands can be used to parse out just the read identifiers into a single file:

```
egrep -o 'HWI[^ ]+' README.md | grep -v "\[" > test_reads.id
```

Next, the full reads associated with the list of identifiers of interest can
be extraced using the biopieces commands:

```
read_fastq -e base_33 -i \
    tcruzi_ystrain/tcruzi-infecting-hsapiens-epimastigote/common/HPGL0254/fastq/HPGL0254_genomic_reads_removed.1.fastq.gz |\
    grab -E test_reads.id -k SEQ_NAME |\
    write_fastq -o test_reads_1.fastq

read_fastq -e base_33 -i \
    tcruzi_ystrain/tcruzi-infecting-hsapiens-epimastigote/common/HPGL0254/fastq/HPGL0254_genomic_reads_removed.2.fastq.gz |\
    grab -E test_reads.id -k SEQ_NAME |\
    write_fastq -o test_reads_2.fastq
```

### Notes

#### Orientation of features within reads

- Based on a quick scan of the raw RNA-Seq reads, there appear to be only about
  13% as many RSL reads as their are SL reads.
- Results are similar for *T. cruzi* / *L. major*
- Table below was generated by searching a single extracellular stage sample 
  each from *T. cruzi* and *L. major*, and counting the number of:
    1. _unanchored_ SL 38-mer suffix hits
    2. _unanchored_ SL 38-mer suffix reverse complement hits
    3. _right-anchored_ 20-mer Poly(A) hits
    4. _left-anchored_ 20-mer Poly(T) hits
- Samples used:
    - HPGL0253 R1 (*T. cruzi*)
    - HPGL0363 R1 (*L. major*)
- Example query used for counting:

```
$ zgrep 'ACTAACGCTATTATTGATACAGTTTCTGTACTATATTG'  HPGL0253/processed/HPGL0253_R1_filtered_trimmed.fastq.gz  | wc -l
```

Feature counts:

|        | T. cruzi| L. major|
|:-------|--------:|--------:|
|SL      |      646|      708|
|RSL     |       88|      107|
|Poly(A) |    52223|    26303|
|Poly(T) |     3202|     1632|

Proportions:

|        | T. cruzi| L. major|
|:-------|--------:|--------:|
|SL      |     0.88|     0.87|
|RSL     |     0.12|     0.13|
|Poly(A) |     0.94|     0.94|
|Poly(T) |     0.06|     0.06|

